packages/core-data/src/resolvers.js

```diff
@@ -23,16 +23,16 @@ import { getKindEntities } from './entities';
  * Requests categories from the REST API, yielding action objects on request
  * progress.
  */
-export async function* getCategories() {
-	const categories = await apiRequest( { path: '/wp/v2/categories?per_page=-1' } );
+export function* getCategories() {
+	const categories = yield apiRequest( { path: '/wp/v2/categories?per_page=-1' } );
```

Comment:
So I'm a bit confused about our intended usage here. In this case, the promise returned by `apiRequest` does not return an action object, yet we're `yield`ing it the same as we would for any other action. I suppose as implemented, it's up to the runtime to determine whether it "looks" like an action to decide whether it should be dispatched? Not always clear at a glance whether the thing being yielded is an action or not.

I guess in my mind, I had imagined something more like:

```
const categories = yield { type: 'API_REQUEST', path: '/wp/v2/categories?per_page=-1' };
```

i.e. always yielding a proper action to which we can attach specific behaviors (through some API to add controls to rungen dynamically). I think this could also help rein in control to prevent abuse of promises by limiting the control flows exposed.

Reply:
> const categories = yield { type: 'API_REQUEST', path: '/wp/v2/categories?per_page=-1' };

I think that's the goal, I won't consider these "actions" but "effects". I didn't want to introduce new effects on the initial implementation.

Also, I tend to think we shouldn't couple the data module to API requests. I'd prefer something more generic:

`yield { type: 'CALL', function: apiRequest, args: { path: '/wp/v2/categories?per_page=-1' } }`

I think it's possible to keep the data module REST API agnostic and still allow you to do what you propose, but this requires having the ability to provide custom controls to the rungen runtime when calling `registerStore`, not sure how I feel about it. I tend to prefer starting small

Reply:
> Also, I tend to think we shouldn't couple the data module to API requests. I'd prefer something more generic:

I know this isn't a democracy, but I agree 😆

Keeping the module API-agnostic means testing/mocking is easier and it's easy to allow people to use their own functions to intercept/modify certain calls if wanted. Win/win in my mind.

Reply:
We've already promoted resolvers as being the swappable layer of abstraction, so do we really have to go the next step further and also make its yielded actions generic?

Reply:
I suppose not, but I find there's a lot of flexibility in this approach.

If I wanted to test the generic approach, it would be easier, for instance.

I suppose I don't see the harm in it but I see benefits.

I'm not sure what the "abuse of promises" would be as mentioned earlier, but I'm generally in favour of us giving ourselves lots of flexibility. I know flexibility is another word for footgun, but I guess I don't see what the danger is here. 🤷‍♂️

Reply:
>I know flexibility is another word for footgun,

This many times.

I've often pointed to the [`effects.js` file](https://github.com/WordPress/gutenberg/blob/master/editor/store/effects.js) as a prime example of flexibility gone wrong, as the file has been a failure in almost every regard: Lack of self-contained documented functions, non-obvious behaviors, callback hell, side effects which aren't side effects, implicit behaviors by reaching into state.

Reply:
For what it's worth though, the proposed, saga-style pattern is more idiomatic and in my experience using something similar, never bit us or made us write confusing/dangerous code on the Add-ons project.

The effects are definitely a bit too out there. I don't think what's being proposed is as flexible or dangerous, and I think perhaps some intense code review around the first set of patches could lead to us creating strong idioms we could document and stick with.

Reply:
holy cannoli.

```js
function* foo() { yield new Promise() }
```

and oh how nasty this is without strong static types,
without the compiler helping make sense of it.

just had to toss that out there :wink: